Add libSyntax support for #tfop and @differentiable.

dan-zheng opened the pull request:
The `-verify-syntax-tree` flag has been enabled when compiling the stdlib:
https://github.com/apple/swift/commit/9cb383edcee217d2cbd3fdc074db417f89fddec9

`-verify-syntax-tree` verifies that no unknown nodes exist in the libSyntax
tree. `#tfop` and `@differentiable` are used in the TensorFlow module of the
stdlib but did not have libSyntax support, causing verification errors during
compilation.

Todo:
- Add libSyntax support for `#adjoint` and `#gradient`.
  - This currently isn't necessary to compile the stdlib because those
    expressions aren't used within the stdlib.

dan-zheng committed:
Add libSyntax support for #tfop and @differentiable.

The `-verify-syntax-tree` flag has been enabled when compiling the stdlib:
https://github.com/apple/swift/commit/9cb383edcee217d2cbd3fdc074db417f89fddec9

`-verify-syntax-tree` verifies that no unknown nodes exist in the libSyntax
tree. `#tfop` and `@differentiable` are used in the TensorFlow module of the
stdlib but did not have libSyntax support, causing verification errors during
compilation.

Todo:
- Add libSyntax support for `#adjoint` and `#gradient`.
  - This currently isn't necessary to compile the stdlib because those
    expressions aren't used within the stdlib.

dan-zheng committed:
Remove unused variable.

dan-zheng committed:
Minor cleanup.

Address comments from @marcrasi.

dan-zheng commented:
Todo before merging: add a round-trip test to `test/Syntax/round_trip_parse_gen.swift`.

nkcsgexi commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -803,14 +843,16 @@ Parser::parseDifferentiableAttribute(SourceLoc atLoc, SourceLoc loc) {
-  auto modeLoc = consumeToken(tok::identifier); // 'forward' or 'reverse'.
+  modeLoc = consumeToken(tok::identifier);
   // Parse comma.
   parseToken(tok::comma, diag::attr_expected_comma, "differentiable",
```
Not sure whether you've seen my comments on the previous PR. Could we use `AttrName` here instead of spelling out "differentiable" ? similarly for the other instances of "differentiable".

dan-zheng replied:
Thanks! Done in https://github.com/apple/swift/pull/17428/commits/913f70d6879a70e55305f90f5d5d2fdf4f593d5e.

marcrasi commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -770,31 +770,71 @@ Parser::parseDifferentiableAttribute(SourceLoc atLoc, SourceLoc loc) {
   ParserStatus Status;
```
This is unused now.

dan-zheng replied:
Thanks! Done in https://github.com/apple/swift/pull/17428/commits/913f70d6879a70e55305f90f5d5d2fdf4f593d5e.

dan-zheng committed:
Minor cleanup.

- Use `AttrName` variable instead of "differentiable" string literal.
- Remove unused variable.

dan-zheng committed:
Add libSyntax tests for #tfop and @differentiable.

dan-zheng commented:
libSyntax round-trip tests for `#tfop` and `@differentiable` added in https://github.com/apple/swift/pull/17428/commits/27bed85fa19a1c8d4535254b77e728f94020505d.

nkcsgexi reviewed (approved):
Thanks for adding the libSyntax tests! Changes LGTM.

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei reviewed (approved):
LGTM
